pkg/frontend/transport/body_seeker.go

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package transport
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+)
+
+// ParseSeekerBodyForm will ensure that request body is a Seeker, parse the form and seek the body back to the beginning.
+func ParseSeekerBodyForm(r *http.Request) error {
+	return withSeekerBody(r, func(r *http.Request) error {
+		return r.ParseForm()
+	})
+}
+
+// withSeekerBody will ensure that request body is a Seeker, call the provided function and seek the body back to the beginning.
+func withSeekerBody(r *http.Request, f func(r *http.Request) error) (err error) {
+	if r.Body == nil {
+		return nil
+	}
+	seeker, ok := r.Body.(io.Seeker)
+	if !ok {
+		seeker, err = newReadCloseSeeker(r.Body)
+		if err != nil {
+			return fmt.Errorf("can't read body: %w", err)
+		}
+		if err := r.Body.Close(); err != nil {
+			return fmt.Errorf("can't close original body: %w", err)
+		}
+	}
```

Comment:
I think we need to set `r.Body` to the seeker at this point, otherwise how could it work?

Reply:
:facepalm: I thought other tests would cover this, but they didn't. Fixed and tested in https://github.com/grafana/mimir/pull/3588/commits/5c9cd1141fc879704e4019479293da1b208e7db8